### PR TITLE
Make events more discoverable and improve way we publish events

### PR DESCRIPTION
**Pull request by tomkerkhove**
Make events more discoverable and improve way we publish events.

- Every custom event now has to inherit from Event and specify the type of the body for the data node. It is also enforced to specify the data version and event type name
- All Azure events now have the data version specified in the event definition and is reflected in the namespace as well
- You know have to pass a full-fledged event and no longer can pass multiple event payloads with only one event id.

This is a breaking change and will be shipped as v2 as we're getting rid of deprecated features.

Relates to #15 & #17 

<details>
<summary>Code Sample</summary>
Example usage of publishing in the new way:

```csharp
var @event = new NewCarRegistered(eventId, eventSubject, licensePlate);

await EventGridPublisherBuilder
    .ForTopic(topicEndpoint)
    .UsingAuthenticationKey(endpointKey)
    .Build()
    .Publish(@event);
```

### CODE REVIEW

File: src/Arcus.EventGrid.EventHubs/Contracts/Events/v1/CaptureFileCreated.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace Arcus.EventGrid.EventHubs.Contracts.Events.v1
+{
+    public class CaptureFileCreated : Event<EventHubCaptureEventData>
+    {
+        public CaptureFileCreated()
+        {
+        }
+
+        public CaptureFileCreated(string id) : base(id)
```
Comment (stijnmoreels) [inline src/Arcus.EventGrid.EventHubs/Contracts/Events/v1/CaptureFileCreated.cs]: Are these valid if there `null` ? If not, should we guard against it?

File: src/Arcus.EventGrid.Publishing/Arcus.EventGrid.Publishing.csproj [hunk 2]
```diff
@@ -19,7 +19,7 @@
     <PackageReference Include="Flurl.Http" Version="2.3.1" />
     <PackageReference Include="Guard.NET" Version="1.2.0" />
     <PackageReference Include="Newtonsoft.Json" Version="11.0.1" />
-    <PackageReference Include="Polly-Signed" Version="5.9.0" />
+    <PackageReference Include="Polly-Signed" Version="5.9.0" />
```
Comment (stijnmoreels) [inline src/Arcus.EventGrid.Publishing/Arcus.EventGrid.Publishing.csproj]: Unnecessary whitespace :)

File: src/Arcus.EventGrid.Publishing/Interfaces/IEventGridPublisherBuilderWithAuthenticationKey.cs [hunk 3]
```diff
@@ -5,6 +5,10 @@ namespace Arcus.EventGrid.Publishing.Interfaces
     /// <summary>
     /// Intermediary builder contract after the <see cref="EventGridPublisherBuilder.ForTopic"/> is called.
     /// </summary>
+    /// <remarks>
+    /// This interface is not explicitly necessary at the moment but could be after there exists another correct way of creating
```
Comment (stijnmoreels) [inline src/Arcus.EventGrid.Publishing/Interfaces/IEventGridPublisherBuilderWithAuthenticationKey.cs]: Yes, indeed. Thanks about that.

File: src/Arcus.EventGrid.Testing/Infrastructure/Hosts/HybridConnectionHost.cs [hunk 4]
```diff
@@ -17,29 +17,21 @@ public class HybridConnectionHost
...
-        /// </summary>
-        public string HostId { get; }
-
         /// <summary>
         ///     Gets the event envelope that includes a requested event (Uses exponential back-off)
         /// </summary>
         /// <param name="eventId">Event id for requested event</param>
         /// <param name="retryCount">Amount of retries while waiting for the event to come in</param>
-        public string GetReceivedEvent(string eventId, int retryCount = 10)
+        public string GetReceivedEvent(string eventId, int retryCount = 5)
```
Comment (stijnmoreels) [inline src/Arcus.EventGrid.Testing/Infrastructure/Hosts/HybridConnectionHost.cs]: Any specific reason why decrease from 10 to 5?

File: src/Arcus.EventGrid.Tests.Unit/Parsing/AzureIoTHubParsingTests.cs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+using Arcus.EventGrid.Parsers;
+using Arcus.EventGrid.Tests.Unit.Artifacts;
+using Xunit;
+
+namespace Arcus.EventGrid.Tests.Unit.Parsing
+{
+    public class AzureIoTHubParsingTests
+    {
+        [Fact]
+        public void Parse_ValidDeviceCreatedEvent_ShouldSucceed2()
```
Comment (stijnmoreels) [inline src/Arcus.EventGrid.Tests.Unit/Parsing/AzureIoTHubParsingTests.cs]: Maybe for maintainability reasons, you can assert on the entire event by (for example) both serializing the expected and actual event.

File: src/Arcus.EventGrid.Tests.Unit/Publishing/EventGridPublisherBuilderTests.cs [hunk 6]
```diff
@@ -25,161 +26,41 @@ public void ForTopic_NullOrEmptyEndpointTopic_ShouldFailWithArgumentException(st
         [InlineData(" ")]
         [InlineData("  ")]
         public void UsingAuthenticationKey_NullOrEmptyEndpointKey_ShouldFailWithArgumentException(string authenticationKey)
-        {
```
Comment (stijnmoreels) [inline src/Arcus.EventGrid.Tests.Unit/Publishing/EventGridPublisherBuilderTests.cs]: Hmm, I think something was wrong merged here. All the tests are removed.

File: src/Arcus.EventGrid.EventHubs/Contracts/Events/v1/CaptureFileCreated.cs (hunk 1 above)
Comment (samneirinck) [reply src/Arcus.EventGrid.EventHubs/Contracts/Events/v1/CaptureFileCreated.cs]: Problem is there's a default constructor already, which allows the consumer to create something with null id's etc, I would assume?

File: src/Arcus.EventGrid.Testing/Infrastructure/Hosts/HybridConnectionHost.cs (hunk 4 above)
Comment (tomkerkhove) [reply src/Arcus.EventGrid.Testing/Infrastructure/Hosts/HybridConnectionHost.cs]: Was done by accident, will revert

File: src/Arcus.EventGrid.EventHubs/Contracts/Events/v1/CaptureFileCreated.cs (hunk 1 above)
Comment (tomkerkhove) [reply src/Arcus.EventGrid.EventHubs/Contracts/Events/v1/CaptureFileCreated.cs]: The base constructor is guarding for empty strings but the default constructors are there for serialization indeed. Not sure if this will work but can try to make them `protected`

File: src/Arcus.EventGrid.Tests.Unit/Parsing/AzureIoTHubParsingTests.cs (hunk 5 above)
Comment (tomkerkhove) [reply src/Arcus.EventGrid.Tests.Unit/Parsing/AzureIoTHubParsingTests.cs]: We already assert on the actual event which is the goal of this test, we can add another test which creates an instance of this event and assert it as well, but would create an issue for that

File: src/Arcus.EventGrid.Tests.Unit/Publishing/EventGridPublisherBuilderTests.cs (hunk 6 above)
Comment (tomkerkhove) [reply src/Arcus.EventGrid.Tests.Unit/Publishing/EventGridPublisherBuilderTests.cs]: Will revert

File: src/Arcus.EventGrid.Publishing/Arcus.EventGrid.Publishing.csproj (hunk 2 above)
Comment (tomkerkhove) [reply src/Arcus.EventGrid.Publishing/Arcus.EventGrid.Publishing.csproj]: Removed them, think the remaining here are just a diff-thing.

image

**Comment by stijnmoreels** [comment]
Seems like there's a lot of formatting changes but not actuall changes. Either we specify exactly what formatting should be used; or maybe these changes should be reverted since they are noise in the actual changes.

**Comment by tomkerkhove** [comment]
That's on the radar to use a tool that aligns everything but it's not there yet due to other priorities. I don't think this should be a blocker for this PR.

Some are documented here: https://dev.azure.com/codit/Arcus/_wiki/wikis/Arcus.wiki?wikiVersion=GBwikiMaster&pagePath=%2FGuidance%2FCoding%20Guidelines

File: src/Arcus.EventGrid.Tests.Core/Events/Data/CarEventData.cs [hunk 7]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Arcus.EventGrid.Tests.Core.Events.Data
+{
+    public class CarEventData
+    {
+        public string LicensePlate { get; set; }
```
Comment (stijnmoreels) [inline src/Arcus.EventGrid.Tests.Core/Events/Data/CarEventData.cs]: Should be made readonly?

File: src/Arcus.EventGrid/EventGridMessage.cs [hunk 8]
```diff
@@ -28,41 +25,6 @@ public EventGridMessage(string sessionId)
...
-            foreach (var eventObject in array.Children<JObject>())
-            {
-                var rawEvent = eventObject.ToString();
-                var gridEvent = JsonConvert.DeserializeObject<Event<TData>>(rawEvent);
-                result.Events.Add(gridEvent);
-            }
-
-            return result;
-        }
+        public List<TEvent> Events { get; } = new List<TEvent>();
```
Comment (stijnmoreels) [inline src/Arcus.EventGrid/EventGridMessage.cs]: Exposing a sequence object entirly in such a matter is a bit bad practice.
If this sequence should not be altered afterwards, expose it as `IEnumerable<>` and call the `.AsEnumerable()` on the list type. (Or it could even be a `Collection` if we only add items).

File: src/Arcus.EventGrid/Parsers/EventGridParser.cs [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+            return eventGridMessage;
+        }
+
+        /// <summary>
+        ///     Parses a string to a EventGridMessage with typed data payload
+        /// </summary>
+        /// <param name="rawJsonBody">Raw JSON body</param>
+        /// <param name="sessionId">Session id for event grid message</param>
+        /// <returns>Typed EventGridMessage</returns>
+        public static EventGridMessage<TEvent> Parse<TEvent>(string rawJsonBody, string sessionId)
```
Comment (stijnmoreels) [inline src/Arcus.EventGrid/Parsers/EventGridParser.cs]: These parameters should also be guarded?

File: src/Arcus.EventGrid.Publishing/EventGridPublisher.cs [hunk 10]
```diff
@@ -53,81 +53,57 @@ internal EventGridPublisher(string topicEndpoint, string authenticationKey, Poli
...
-        /// <typeparam name="TData">Type of the specific EventData</typeparam>
-        /// <param name="subject">Subject</param>
-        /// <param name="eventType">Event Type</param>
-        /// <param name="data">Typed event data</param>
-        /// <param name="id">Unique id of the event, generated by default</param>
-        /// <returns></returns>
-        public async Task Publish<TData>(string subject, string eventType, IEnumerable<TData> data, string id = null)
+        /// <typeparam name="TEvent">Type of the specific EventData</typeparam>
+        /// <param name="events">Events to publish</param>
+        public async Task Publish<TEvent>(List<TEvent> events)
```
Comment (samneirinck) [inline src/Arcus.EventGrid.Publishing/EventGridPublisher.cs]: Is there a reason this is a `List<TEvent>`? (similar for other overloads)

```suggestion
        public async Task Publish<TEvent>(IEnumerable<TEvent> events)
```

File: src/Arcus.EventGrid.Tests.Core/Events/Data/CarEventData.cs (hunk 7 above)
Comment (samneirinck) [reply src/Arcus.EventGrid.Tests.Core/Events/Data/CarEventData.cs]: I fear it might need a setter for (de)serialization. It always comes down to that default constructor 😨

Comment (tomkerkhove) [reply src/Arcus.EventGrid.Tests.Core/Events/Data/CarEventData.cs]: And it's just a test class as well 🤷‍♂️

File: src/Arcus.EventGrid.Publishing/EventGridPublisher.cs (hunk 10 above)
Comment (tomkerkhove) [reply src/Arcus.EventGrid.Publishing/EventGridPublisher.cs]: Not really, just a (bad) habit

File: src/Arcus.EventGrid/Parsers/EventGridParser.cs (hunk 9 above)
Comment (tomkerkhove) [reply src/Arcus.EventGrid/Parsers/EventGridParser.cs]: Good point, adding it along with tests

File: src/Arcus.EventGrid.Tests.Core/Events/Data/CarEventData.cs (hunk 7 above)
Comment (stijnmoreels) [reply src/Arcus.EventGrid.Tests.Core/Events/Data/CarEventData.cs]: Sigh... *thinks about F#*

File: src/Arcus.EventGrid.Publishing/EventGridPublisher.cs (hunk 10 above)
Comment (tomkerkhove) [reply src/Arcus.EventGrid.Publishing/EventGridPublisher.cs]: So this is kinda funny. If I switch this to IEnumerable it thinks that List<TEvent> is a single event and it cannot cast it to TEvent.

I would suggest to either keep it as is or rename `Publish<TEvent>(IEnumerable<TEvent> events)` to `PublishMany<TEvent>(IEnumerable<TEvent> events)`

File: src/Arcus.EventGrid/EventGridMessage.cs (hunk 8 above)
Comment (stijnmoreels) [reply src/Arcus.EventGrid/EventGridMessage.cs]: Yeah, if you feel more comfortable with that; could also work; but either way without exposing the manipulation of the sequence. In the context of "Only expose what you need".

File: src/Arcus.EventGrid.Tests.Core/Events/Data/CarEventData.cs (hunk 7 above)
Comment (tomkerkhove) [reply src/Arcus.EventGrid.Tests.Core/Events/Data/CarEventData.cs]: Don't see the problem with this to be honest :)

File: src/Arcus.EventGrid/EventGridMessage.cs (hunk 8 above)
Comment (tomkerkhove) [reply src/Arcus.EventGrid/EventGridMessage.cs]: Ok for you @samneirinck?

File: src/Arcus.EventGrid.Publishing/EventGridPublisher.cs (hunk 10 above)
Comment (tomkerkhove) [reply src/Arcus.EventGrid.Publishing/EventGridPublisher.cs]: Thoughts @samneirinck/@stijnmoreels?

Comment (samneirinck) [reply src/Arcus.EventGrid.Publishing/EventGridPublisher.cs]: Prefer option 2, since if I pass in an `IEnumerable<T>`/`T[]`/,,,, I would not get the behavior I expected.

File: src/Arcus.EventGrid.Tests.Unit/Publishing/EventGridPublisherTests.cs [hunk 11]
```diff
@@ -2,141 +2,65 @@
...
-            };
+            NewCarRegistered @event = null;
 
             // Act
-            var eventGridPublisher = 
+            var eventGridPublisher =
                 EventGridPublisherBuilder
                     .ForTopic(topicEndpoint)
                     .UsingAuthenticationKey(authenticationKey)
                     .Build();
```
Comment (jornbeyers) [inline src/Arcus.EventGrid.Tests.Unit/Publishing/EventGridPublisherTests.cs]: as you have Arrange and Act comments, add some Assert comments as well

File: src/Arcus.EventGrid.Tests.Unit/Publishing/EventGridPublisherBuilderTests.cs [hunk 12]
```diff
@@ -160,10 +155,10 @@ public void WithExponentialRetryWithCircuitBroker_PositiveExceptionsAllowedBefor
         {
             // Arrange
             const int positiveInt = 10;
-            TimeSpan positiveInterval = TimeSpan.FromSeconds(5);
+            var positiveInterval = TimeSpan.FromSeconds(5);
 
             // Act
-            IEventGridPublisher publisher =
+            var publisher =
                 CreateEventGridBuilder()
```
Comment (jornbeyers) [inline src/Arcus.EventGrid.Tests.Unit/Publishing/EventGridPublisherBuilderTests.cs]: is it the diff or is this accidently moved to a newline?

File: src/Arcus.EventGrid.Tests.Unit/Parsing/EventGridParsingTests.cs [hunk 13]
```diff
@@ -36,5 +37,100 @@ public void Parse_ValidSubscriptionValidationEvent_ShouldSucceed()
             Assert.NotNull(eventPayload.Data);
             Assert.Equal(validationCode, eventPayload.Data.ValidationCode);
         }
+
+        [Fact]
+        public void Parse_ValidSubscriptionValidationEventWithSessionId_ShouldSucceed()
+        {
+            // Arrange
+            string rawEvent = EventSamples.SubscriptionValidationEvent;
```
Comment (jornbeyers) [inline src/Arcus.EventGrid.Tests.Unit/Parsing/EventGridParsingTests.cs]: Just curious: what are the guidelines regarding the usage of var?

File: src/Arcus.EventGrid.Tests.Unit/Events/EventCreationTests.cs [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+            string eventId = Guid.NewGuid().ToString();
+            const string licensePlate = "1-TOM-337";
+
+            // Act
+            var createdEvent = new NewCarRegistered(eventId, licensePlate);
+            Assert.NotNull(createdEvent);
+            Assert.Equal(eventId, createdEvent.Id);
+            Assert.Null(createdEvent.Subject);
+            Assert.NotNull(createdEvent.Data);
+            Assert.Equal(licensePlate, createdEvent.Data.LicensePlate);
```
Comment (jornbeyers) [inline src/Arcus.EventGrid.Tests.Unit/Events/EventCreationTests.cs]: Some missing // Assert comments

File: src/Arcus.EventGrid.Tests.Unit/Parsing/EventGridParsingTests.cs (hunk 13 above)
Comment (tomkerkhove) [reply src/Arcus.EventGrid.Tests.Unit/Parsing/EventGridParsingTests.cs]: See https://github.com/arcus-azure/arcus.eventgrid/pull/36#issuecomment-432599967

WIP but will be enforced in the future.

File: src/Arcus.EventGrid.Tests.Unit/Publishing/EventGridPublisherBuilderTests.cs (hunk 12 above)
Comment (tomkerkhove) [reply src/Arcus.EventGrid.Tests.Unit/Publishing/EventGridPublisherBuilderTests.cs]: It's the diff

File: src/Arcus.EventGrid.Publishing/EventGridPublisher.cs (hunk 10 above)
Comment (tomkerkhove) [reply src/Arcus.EventGrid.Publishing/EventGridPublisher.cs]: Agreed, changed this

File: src/Arcus.EventGrid/EventGridMessage.cs (hunk 8 above)
Comment (stijnmoreels) [reply src/Arcus.EventGrid/EventGridMessage.cs]: Now, only this and it's OK I think.

File: src/Arcus.EventGrid.EventHubs/Contracts/Events/v1/CaptureFileCreated.cs (hunk 1 above)
Comment (stijnmoreels) [reply src/Arcus.EventGrid.EventHubs/Contracts/Events/v1/CaptureFileCreated.cs]: Ah, there exists some other event types that also needs guarding I think?

File: src/Arcus.EventGrid/EventGridMessage.cs (hunk 8 above)
Comment (tomkerkhove) [reply src/Arcus.EventGrid/EventGridMessage.cs]: Done!

**Comment by tomkerkhove** [comment]
I think everything is covered, feel free to do a final sweep @stijnmoreels 	@samneirinck 	@jornbeyers

File: src/Arcus.EventGrid.Tests.Unit/Publishing/EventGridPublisherBuilderTests.cs [hunk 15]
```diff
@@ -73,22 +70,20 @@ public void WithExponentialRetry_NegativeRetryCount_ShouldFailWithArgumentOutOfR
...
             // Arrange
-            TimeSpan positiveInterval = TimeSpan.FromSeconds(5);
+            var positiveInterval = TimeSpan.FromSeconds(5);
 
             // Act
-            IEventGridPublisher publisher =
-                CreateEventGridBuilder()
-                    .WithCircuitBreaker<Exception>(10, positiveInterval)
-                    .Build();
+            var publisher =CreateEventGridBuilder()
```
Comment (stijnmoreels) [inline src/Arcus.EventGrid.Tests.Unit/Publishing/EventGridPublisherBuilderTests.cs]: :) forgot a space.

Comment (tomkerkhove) [reply src/Arcus.EventGrid.Tests.Unit/Publishing/EventGridPublisherBuilderTests.cs]: Our code styling tooling will fix this in the future, see #39

File: src/Arcus.EventGrid/EventGridMessage.cs [hunk 16]
```diff
@@ -1,23 +1,24 @@
...
-    public class EventGridMessage<TData>
+    public class EventGridMessage<TEvent> where TEvent : IEvent, new()
     {
         /// <summary>
         /// Creates event grid message with passed session Id
         /// </summary>
         /// <param name="sessionId">Unique session id for all batched messages</param>
-        public EventGridMessage(string sessionId)
+        /// <param name="events">List of events that are part of this Event Grid message</param>
+        public EventGridMessage(string sessionId, IList<TEvent> events)
```
[the remark this answers is not in the document]
Comment (tomkerkhove) [reply src/Arcus.EventGrid/EventGridMessage.cs]: [ReadOnlyCollection<TEvent>](https://docs.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.readonlycollection-1?redirectedfrom=MSDN&view=netframework-4.7.2#constructors) constructor only supports `IList<TEvent>`

**Comment by tomkerkhove** [comment]
Thank you @samneirinck 	@stijnmoreels 	@jornbeyers 	for the review!
